src/codegen/compile.ml

```diff
@@ -9112,6 +8945,47 @@ module FuncDec = struct
     | _ -> ()
     end
 
+  let async_stabilization_method_name = "__motoko_async_stabilization"
+
+  let call_async_stabilization env =
+    (* TODO: Avoid temporary blob allocation for self reference *)
+    IC.get_self_reference env ^^ Blob.as_ptr_len env ^^
+    Blob.lit_ptr_len env async_stabilization_method_name ^^
+    ignoring_callback env ^^ compile_unboxed_zero ^^
+    (* TODO: Check usability: Traps in this async callee are not displayed in drun/dfx but only debug outputs *)
+    ignoring_callback env ^^ compile_unboxed_zero ^^
+    IC.system_call env "call_new" ^^
+    IC.system_call env "call_perform" ^^
+    E.then_trap_with env "Async stabilization increment call failed"
+
+  let export_async_stabilization_method env =
+    let name = async_stabilization_method_name in
+    begin match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.define_built_in env name [] [] (fun env ->
+        (* All messages are blocked except this method and the upgrade *)
+        Lifecycle.trans env Lifecycle.InStabilization ^^
+        IC.assert_caller_self_or_controller env ^^
+        (* Skip argument deserialization to avoid allocations *)
+        GraphCopyStabilization.stabilization_increment env ^^
+        (G.if0
+          G.nop
+          (call_async_stabilization env)) ^^
+        (* Send static reply *)
+        Blob.lit_ptr_len env "DIDL\x00\x00" ^^
+        IC.reply_with_data env
+        (* Skip garbage collection *)
```

Comment:
I think this should just exit, with replying. Instead, the call_back should test if there is more work to be done and either do it and issue another call with same callback and exit, or (if done), issue a reply and exit. Not sure about the error handling but it could probably just reject with the error message and exit.

Reply:
Thank you for the suggestion. I implemented it in this way.